botocore/signers.py

```diff
@@ -188,31 +195,49 @@ def generate_presigned_url(self, request_dict, expires_in=3600,
         """
         if region_name is None:
             region_name = self._region_name
-        query_prefix = '-query'
+        query_suffix = '-query'
         signature_version = self._signature_version
-        if not signature_version.endswith(query_prefix):
-            signature_version += query_prefix
+        if signature_version is not botocore.UNSIGNED and not \
+                signature_version.endswith(query_suffix):
+            signature_version += query_suffix
 
-        kwargs = {'signing_name': self._signing_name,
-                  'region_name': region_name,
-                  'signature_version': signature_version,
-                  'expires': expires_in}
+        handler, response = self._event_emitter.emit_until_response(
+            'choose-signer.{0}.{1}'.format(self._service_name, operation_name),
+            signing_name=self._signing_name, region_name=self._region_name,
+            signature_version=signature_version)
 
-        signature_type = signature_version.split('-', 1)[0]
-        try:
-            auth = self.get_auth_instance(**kwargs)
-        except UnknownSignatureVersionError:
-            raise UnsupportedSignatureVersionError(
-                signature_version=signature_type)
+        if response is not None:
+            signature_version = response
+
+        if signature_version is botocore.UNSIGNED:
+            signature_type = botocore.UNSIGNED
+        else:
+            # The suffix needs to be checked again in case we got an improper
+            # signature version from choose-signer.
+            if not signature_version.endswith(query_suffix):
+                signature_version += query_suffix
+            signature_type = signature_version.split('-', 1)[0]
 
         request = create_request_object(request_dict)
 
         # Fix s3 host for s3 sigv2 bucket names
         fix_s3_host(request, signature_type, region_name)
 
-        auth.add_auth(request)
-        request.prepare()
+        if signature_type is not botocore.UNSIGNED:
```

Comment:
What are your thoughts on consolidating ``signature_type`` and ``signature_version`` and just use ``signature_version``? Looks like ``signature_type`` was only used for error messaging. This would help reduce all of the if else statements throughout the code.